### PR TITLE
RavenDB-6103: Reuse journal files

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 1]
```diff
@@ -208,26 +231,51 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
...
+                    throw new InvalidDataException("Committed and not empty transaction hash can't be equal to 0");
+                if (current->CompressedSize <= 0)
+                    throw new InvalidDataException("Compression error in transaction.");
+                
+                return ActionOnInvalidTxId.FailOnInvalidHash;
+
+            }
+            else
+            {
+                // TODO : what to do on TxId == 1 ??
```
Comment (aviviadi) [inline src/Voron/Impl/Journal/JournalReader.cs:276]: @ayende Should we fail here ?

File: test/FastTests/Blittable/MemoryPoolTests.cs [hunk 2]
```diff
@@ -1,4 +1,5 @@
-﻿using System.Collections.Concurrent;
```
Comment (aviviadi) [inline test/FastTests/Blittable/MemoryPoolTests.cs:1]: please ignore this file

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 3]
```diff
@@ -65,8 +66,19 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options, bool
...
+            if (checkCrc)
+            {
+                if (!ValidatePagesHash(options, current))
+                {
+                    return false;
+                }
+                if (actionOnInvalidTxId == ActionOnInvalidTxId.FailOnValidHash)
+                {
+                    throw new InvalidDataException(
+                        "Page has valid(!) hash but transaction id bigger more then '1' from last tx. transaction id = " +
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs:78]: Don't do `more than '1'` user has no concept, add the actual number

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 4]
```diff
@@ -161,28 +174,38 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
             current = (TransactionHeader*)
                 (_journalPager.AcquirePagePointer(this, pageNumber) + positionInsidePage);
 
+            // due to the reuse of journals we no longer can assume we have zeros in the end of the journal
+            // we might have there random garbage or old transactions we can ignore, so we have the following scenarios:
+            // * TxId <= current Id      ::  we can ignore old transaction of the reused journal and continue
+            // * TxId == current Id + 1  ::  valid, but if hash is invalid - we can assume corruption
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs:180]: Not corruption, just that the transaction hasn't been committed

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 5]
```diff
@@ -208,26 +231,51 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
...
+                if (txIdDiff > 1)
+                    return ActionOnInvalidTxId.FailOnValidHash; // if hash valid(!) and we're skipping TXs then something might be wrong
+
+                // if (txIdDiff == 1) :
+                if (current->TxMarker.HasFlag(TransactionMarker.Commit) && current->LastPageNumber < 0)
+                    throw new InvalidDataException("Last page number after committed transaction must be greater than 0");
+                if (current->TxMarker.HasFlag(TransactionMarker.Commit) && current->PageCount > 0 && current->Hash == 0)
+                    throw new InvalidDataException("Committed and not empty transaction hash can't be equal to 0");
+                if (current->CompressedSize <= 0)
+                    throw new InvalidDataException("Compression error in transaction.");
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs:269]: Need better error msg here

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 6]
```diff
@@ -1163,7 +1163,7 @@ private CompressedPagesResult PrepareToWriteToJournal(LowLevelTransaction tx, Ab
             txHeader->CompressedSize = compressedLen;
             txHeader->UncompressedSize = totalSizeWritten;
             txHeader->PageCount = numberOfPages;
-            txHeader->Hash = Hashing.XXHash64.Calculate(compressionBuffer, (ulong)compressedLen);
+            txHeader->Hash = Hashing.XXHash64.Calculate(compressionBuffer, (ulong)compressedLen, (ulong)txHeader->TransactionId);
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs:1166]: Need to be sure to update the voron format here.
`Voron.Global.Constants.CurrentVersion`

File: src/Voron/StorageEnvironmentOptions.cs [hunk 7]
```diff
@@ -25,7 +27,7 @@ public abstract class StorageEnvironmentOptions : IDisposable
 
         public event EventHandler<RecoveryErrorEventArgs> OnRecoveryError;
         public event EventHandler<NonDurabalitySupportEventArgs> OnNonDurabaleFileSystemError;
-
+        private static long _reuseCounter;
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:30]: why static?

File: src/Voron/StorageEnvironmentOptions.cs [hunk 8]
```diff
@@ -190,6 +196,42 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
                     return new Win32MemoryMapPager(this, FilePath, InitialFileSize, usePageProtection: true);
                     //return new SparseMemoryMappedPager(this, FilePath, InitialFileSize);
                 });
+
+                var dirInfo = new DirectoryInfo(_journalPath);
+                var journalsForReuse = dirInfo.GetFiles("*.pending-recycle").Select(x => x.Name).ToList();
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:201]: Why call `ToList` if you are just going to iterate and throw in a bit?

File: src/Voron/StorageEnvironmentOptions.cs [hunk 9]
```diff
@@ -190,6 +196,42 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
...
+                var fileInfoList = new List<Tuple<string, DateTime>>();
+                foreach (var reuseFileNameIter in journalsForReuse)
+                {
+                    try
+                    {
+                        var reuseNameWithoutExt = Path.GetFileNameWithoutExtension(reuseFileNameIter);
+                        var reuseNum = long.Parse(reuseNameWithoutExt);
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:208]: TryParse

File: src/Voron/StorageEnvironmentOptions.cs [hunk 10]
```diff
@@ -190,6 +196,42 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
...
+                        if (reuseNum > _reuseCounter)
+                            _reuseCounter = reuseNum;
+                    }
+                    catch
+                    {
+                        // ignore
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:214]: DeleteFile?

File: src/Voron/StorageEnvironmentOptions.cs [hunk 11]
```diff
@@ -190,6 +196,42 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
                     return new Win32MemoryMapPager(this, FilePath, InitialFileSize, usePageProtection: true);
                     //return new SparseMemoryMappedPager(this, FilePath, InitialFileSize);
                 });
+
+                var dirInfo = new DirectoryInfo(_journalPath);
+                var journalsForReuse = dirInfo.GetFiles("*.pending-recycle").Select(x => x.Name).ToList();
+                var fileInfoList = new List<Tuple<string, DateTime>>();
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:202]: Use SortedList here, instead.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 12]
```diff
@@ -250,6 +292,79 @@ public override IJournalWriter CreateJournalWriter(long journalNumber, long jour
...
+                                " times and then create regular new journal #" +
+                                journalNumber, ex);
+                        if (retries < 1)
+                            return null;
+                    }
+                }
+
+                if (reused == null)
+                {
+                    if (_log.IsInfoEnabled)
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:330]: Why is this logged? we will always have it when we start, at least

File: src/Voron/StorageEnvironmentOptions.cs [hunk 13]
```diff
@@ -250,6 +292,79 @@ public override IJournalWriter CreateJournalWriter(long journalNumber, long jour
...
+                        _log.Info(
+                            "Failed to get available journal for reuse (might be deleted by the user while server is up, which is a valid operation). Will generate regular new journal file");
+                    return null;
+                }
+
+                // check if needed to remove very old pending recycle files:
+                foreach (var elementExist in _journalsForReuse)
+                {
+                    if (DateTime.UtcNow - elementExist.Item2 > TimeSpan.FromHours(72))
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:339]: Wrong, this should be 72 hours from the newest element in here.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 14]
```diff
@@ -557,6 +679,9 @@ public static bool RunningOnPosix
         public OpenFlags SafePosixOpenFlags = OpenFlags.O_DSYNC | PerPlatformValues.OpenFlags.O_DIRECT;
         private readonly Logger _log;
 
+        private readonly ConcurrentStack<Tuple<string, DateTime>> _journalsForReuse =
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:682]: This doesn't need to be concurrent, and better be SortedList

File: src/Voron/StorageEnvironmentOptions.cs [hunk 15]
```diff
@@ -190,6 +196,42 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
...
+                var journalsForReuse = dirInfo.GetFiles("*.pending-recycle").Select(x => x.Name).ToList();
+                var fileInfoList = new List<Tuple<string, DateTime>>();
+                foreach (var reuseFileNameIter in journalsForReuse)
+                {
+                    try
+                    {
+                        var reuseNameWithoutExt = Path.GetFileNameWithoutExtension(reuseFileNameIter);
+                        var reuseNum = long.Parse(reuseNameWithoutExt);
+                        if (reuseNum > _reuseCounter)
+                            _reuseCounter = reuseNum;
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:210]: Not thread safe here, and shouldn't be, make this instance value

File: src/Voron/StorageEnvironmentOptions.cs [hunk 16]
```diff
@@ -571,5 +696,39 @@ public virtual void SetPosixOptions()
...
         }
+
+        public bool TryStoreJournalForReuse(string filename)
+        {
+            bool success = true;
+            try
+            {
+                var fileModifiedDate = new FileInfo(filename).LastWriteTimeUtc;
+                var reuseCounter = Interlocked.Increment(ref _reuseCounter);
+                var newName = Path.GetDirectoryName(filename) + Path.DirectorySeparatorChar + PendingRecycleName(reuseCounter);
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:707]: Never do that, use `Path.Combine`

File: test/FastTests/Blittable/MemoryPoolTests.cs (hunk 2 above)
Comment (ayende) [reply test/FastTests/Blittable/MemoryPoolTests.cs:1]: so why is it here

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/StorageEnvironmentOptions.cs (hunk 12 above)
Comment (aviviadi) [reply src/Voron/StorageEnvironmentOptions.cs:330]: No it isn't.  We TryPop from _journalsForReuse and if it is empty (at start i.e.) then we returning null from this method

File: src/Voron/Impl/Journal/JournalReader.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/JournalReader.cs:78]: Name `FailOnValidHash` is confusing. It doesn't state that `txIdDiff > 1`.

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 17]
```diff
@@ -161,28 +174,38 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
...
+                // if the header marker is zero or garbage, we are probably in the area at the end of the log file, and have no additional log records
                 // to read from it. This can happen if the next transaction was too big to fit in the current log file. We stop reading
-                // this log file and move to the next one. 
+                // this log file and move to the next one, or it might have happened becuase of reuse of journal file 
 
-                RequireHeaderUpdate = current->HeaderMarker != 0;
-                if (RequireHeaderUpdate)
-                {
-                    options.InvokeRecoveryError(this, "Transaction " + current->TransactionId + " header marker was set to garbage value, file is probably corrupted", null);
-                }
```
Comment (arekpalinski) [inline src/Voron/Impl/Journal/JournalReader.cs:176]: Didn't we loose the validation check and raising recovery error here? I don't see `options.InvokeRecoveryError` to be moved somewhere.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 18]
```diff
@@ -571,5 +686,39 @@ public virtual void SetPosixOptions()
...
+
+                File.Move(filename, newName);
+                _journalsForReuse.Push(new Tuple<string, DateTime>(newName, fileModifiedDate));
```
Comment (arekpalinski) [inline src/Voron/StorageEnvironmentOptions.cs:710]: Is is ok to not have any size limit on `_journalsForReuse`?

File: src/Voron/StorageEnvironmentOptions.cs [hunk 19]
```diff
@@ -190,6 +196,32 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
...
                     //return new SparseMemoryMappedPager(this, FilePath, InitialFileSize);
                 });
+
+                var dirInfo = new DirectoryInfo(_journalPath);
+                var journalsForReuse = dirInfo.GetFiles("*.pending-recycle").Select(x => x.Name).ToList();
+
+                var fileInfoList = new List<Tuple<string, DateTime>>();
+                foreach (var reuseFileNameIter in journalsForReuse)
+                {
+                    var reuseFileName = dirInfo.ToString() + reuseFileNameIter;
```
Comment (arekpalinski) [inline src/Voron/StorageEnvironmentOptions.cs]: `Path.Combine`

File: src/Voron/StorageEnvironmentOptions.cs [hunk 20]
```diff
@@ -199,11 +199,21 @@ public DirectoryStorageEnvironmentOptions(string basePath, string tempPath, stri
...
+                        var reuseNameWithoutExt = Path.GetFileNameWithoutExtension(reuseFileNameIter);
+                        var reuseNum = long.Parse(reuseNameWithoutExt);
+                        if (reuseNum > _reuseCounter)
+                            _reuseCounter = reuseNum;
+                    }
+                    catch
+                    {
+                        // ignore
+                    }
+                    var reuseFileName = dirInfo + reuseFileNameIter;
```
Comment (arekpalinski) [inline src/Voron/StorageEnvironmentOptions.cs:216]: Path.Combine

**Comment by arekpalinski** [comment]
Once we run this on indexes of SO dataset we will end up with dozen of journals residing on the disk even though the indexing work is done.

We might want to cleanup the files more frequently than 72 hours.

**Comment by arekpalinski** [comment]
Note that we already have `StorageEnvironment.Cleanup`

**Comment by arekpalinski** [comment]
It would be good to have some tests for journals reuse / cleanup

**Pull request closed without merging**